In [1]:
from __future__ import print_function
import sklearn
from sklearn.feature_extraction.text import TfidfVectorizer
from sklearn.feature_extraction.text import CountVectorizer
from sklearn.decomposition import TruncatedSVD
from sklearn.preprocessing import Normalizer
from sklearn import metrics
from sklearn.cluster import KMeans, MiniBatchKMeans
import pandas as pd
import numpy
import numpy as np
import matplotlib.pyplot as plt
import os
import io
import re

In [2]:
import sklearn
from nltk.classify.scikitlearn import  SklearnClassifier
from sklearn.svm import SVC, LinearSVC,  NuSVC
from sklearn.naive_bayes import  MultinomialNB, BernoulliNB
from sklearn.linear_model import  LogisticRegression
from sklearn.metrics import  accuracy_score

In [7]:
pip install nltk

Note: you may need to restart the kernel to use updated packages.


In [9]:
import nltk
from nltk.collocations import  BigramCollocationFinder
from nltk.metrics import  BigramAssocMeasures

In [10]:
from nltk.probability import  FreqDist,ConditionalFreqDist
from nltk.metrics import  BigramAssocMeasures

In [4]:
import pandas as pd
path='c:/Python/'
df=pd.read_excel(path+"天成情緒文本.xlsx",index_col=None)
df

text tag
0                              地點離捷運站近，要去台北景點都方便服務人員熱心   P
1                          有問題也都能即時解答房間乾淨明亮，設備俱全，很安靜舒適   P
2                                   這次已經是第二次入住，服務比上次更佳   P
3                                        職員對旅客有賓至如歸的感覺   P
4                 很寬敞，很乾凈，也很安靜。我們住的很舒服，也很放鬆。酒店的早餐也很不錯。   P
5                                        離火車站非常近，交通又便利   P
6                                             櫃台人員態度親切   P
7                                   內部裝潢簡單乾淨，舒適悠閒，值得好推   P
8                                          服務得週到，房間很整潔   P
9                                            早餐不錯 東西好吃   P
10                         房間內設施一應俱全、非常好的一次旅行體驗，下次還要光臨   P
11                            很氣派，很豪華，內部裝飾保養很好，還有自動馬桶蓋   P
12                                         高CP值&服務態度佳!   P
13                                        服務人員都很用心解答問題   P
14                                   旁邊有便利商店。有免費wifi使用   P
15                             房間十分乾淨，客服也很有禮貌及體貼，值得推薦。   P
16                                           衛生間的馬桶很好用   P
17                             飯店大堂設有台北觀光小冊子，以便入住的旅客查閱   P
18                                 位置極好，樓下就是 7-11，十分方便   P
19                             麻雀雖小五臟俱全，像回到家的親切，舒適安靜。    P
20                 服務及接待人員像家人般的溫馨，飯店地點在台北車站旁交通方便，四通八達。   P
21                                    浴室設備非常自動化；早餐款式多。   P
22                           服務人員的熱情都是有目共睹的。我們全家都覺得非常棒   P
23                       飯店附近有百貨公司，地下街商店，購物逛街 用餐都相當便利。   P
24                            前台服務也很週到，有禮貌，下次來台考慮再入住天成   P
25   服務員有禮貌, 房間整潔舒適, 浴室浴品齊備, 淋浴水力足夠, 電視頻道多元化, 節目豐富,...   P
26                        附近購物地點多，也有很多景點遊覧。 價錢適中，接待十分好   P
27                                    房間翻新過整齊清潔。員工友善熱誠   P
28                                房間很整潔，服務員很有禮貌，熱水即開即熱   P
29                         喜歡他們的多功能設備化，例如全燈一鍵控制；自動洗廁機等   P
..                                                 ...  ..
90                   洗澡的水量有點小，洗澡時要沖很久才會覺得乾淨，還有沐浴乳也不太好沖   N
91                              櫃臺服務人員提供相當清楚的資訊讓我們打發時間   N
92   雖然房間浴室都稍小，還沒有免費的自助洗衣跟早餐，房價也稍不美麗，但看在它的金地段，這一切都可...   N
93                                          小孩設備考慮不夠週到   N
94                                            酒店的大樓比較老   N
95                        可惜美中不足的是浴缸裡有頭髮，可能需要再加強一下房間清潔   N
96                            房間內無線網路不穩定，常常有2格訊號但卻無法上網   N
97                                           不過二人房間偏小。   N
98                        惟酒店較舊，沒有企身浴缸，浴缸太高，不適合同行小童及長者   N
99             房间光线可以再亮一些，另外，淋浴热水需要调节冷热两个旋钮，感觉有点不方便。     N
100                      只能开窗子，然而窗子又小，風都不怎么能吹进来又有很大的噪音   N
101                                         缺点是充电插座太少。   N
102                                         床及枕頭太軟不太習慣   N
103                  设备老旧，尤其卫生间内没有独立淋浴，需要爬进浴缸，不适合行动不便者   N
104                                      服务有待提升，无线很不稳定   N
105                                隔音加強，減少干擾，停車空間，稍嫌不足   N
106                                       空調差了點 睡一晚都是塵   N
107                                      酒店設施比較舊，尤其是廁所   N
108                                        電視機的接收效果也不好   N
109                      看到我們是背包客,前台服務員甚至懶得接待我們,自顧自的做事   N
110                        房間並不像之前評論里說的那樣好,也不像酒店網站上的照片   N
111                                      房間建議是讓電源插座靠近床   N
112                                          自助早餐不是很壯觀   N
113                                    酒店本身很舊需要重新裝修一下了   N
114                                            不要指望游泳池   N
115                                             裝飾有點過時   N
116                                     漏水和一個吹風機還是在地面上   N
117                                     這絕對根本不應該發生這樣的事   N
118                                            房間很小很擁擠   N
119                                      有點怪味兒的健身房設施有限   N

[120 rows x 2 columns]

In [5]:
import jieba
import jieba.analyse
from optparse import OptionParser
import re

# 設定斷詞字典  
jieba.set_dictionary(path + "dict.txt.big.txt")    
# 呼叫自行定義的詞典(增加詞彙) 
jieba.load_userdict(path + "userdict.txt")    

# 結巴分詞
token=[]
for n in range(len(df)):
    content=df.text[n]
    result = jieba.cut(content)
    # 分詞後存成 string，中間空白隔開
    tok=' '.join(result)
    token.append(tok)

Building prefix dict from c:\Python\dict.txt.big.txt ...
Loading model from cache C:\Users\loy07\AppData\Local\Temp\jieba.uba39d21d37c399eb7eb7fd795ed896e7.cache
Loading model cost 1.931 seconds.
Prefix dict has been built succesfully.


In [8]:
df['token']=token
df

text tag  \
0                              地點離捷運站近，要去台北景點都方便服務人員熱心   P   
1                          有問題也都能即時解答房間乾淨明亮，設備俱全，很安靜舒適   P   
2                                   這次已經是第二次入住，服務比上次更佳   P   
3                                        職員對旅客有賓至如歸的感覺   P   
4                 很寬敞，很乾凈，也很安靜。我們住的很舒服，也很放鬆。酒店的早餐也很不錯。   P   
5                                        離火車站非常近，交通又便利   P   
6                                             櫃台人員態度親切   P   
7                                   內部裝潢簡單乾淨，舒適悠閒，值得好推   P   
8                                          服務得週到，房間很整潔   P   
9                                            早餐不錯 東西好吃   P   
10                         房間內設施一應俱全、非常好的一次旅行體驗，下次還要光臨   P   
11                            很氣派，很豪華，內部裝飾保養很好，還有自動馬桶蓋   P   
12                                         高CP值&服務態度佳!   P   
13                                        服務人員都很用心解答問題   P   
14                                   旁邊有便利商店。有免費wifi使用   P   
15                             房間十分乾淨，客服也很有禮貌及體貼，值得推薦。   P   
16                                           衛生間的馬桶很好用   P   
17                             飯店大堂設有台北觀光小冊子，以便入住的旅客查閱   P   
18                                 位置極好，樓下就是 7-11，十分方便   P   
19                             麻雀雖小五臟俱全，像回到家的親切，舒適安靜。    P   
20                 服務及接待人員像家人般的溫馨，飯店地點在台北車站旁交通方便，四通八達。   P   
21                                    浴室設備非常自動化；早餐款式多。   P   
22                           服務人員的熱情都是有目共睹的。我們全家都覺得非常棒   P   
23                       飯店附近有百貨公司，地下街商店，購物逛街 用餐都相當便利。   P   
24                            前台服務也很週到，有禮貌，下次來台考慮再入住天成   P   
25   服務員有禮貌, 房間整潔舒適, 浴室浴品齊備, 淋浴水力足夠, 電視頻道多元化, 節目豐富,...   P   
26                        附近購物地點多，也有很多景點遊覧。 價錢適中，接待十分好   P   
27                                    房間翻新過整齊清潔。員工友善熱誠   P   
28                                房間很整潔，服務員很有禮貌，熱水即開即熱   P   
29                         喜歡他們的多功能設備化，例如全燈一鍵控制；自動洗廁機等   P   
..                                                 ...  ..   
90                   洗澡的水量有點小，洗澡時要沖很久才會覺得乾淨，還有沐浴乳也不太好沖   N   
91                              櫃臺服務人員提供相當清楚的資訊讓我們打發時間   N   
92   雖然房間浴室都稍小，還沒有免費的自助洗衣跟早餐，房價也稍不美麗，但看在它的金地段，這一切都可...   N   
93                                          小孩設備考慮不夠週到   N   
94                                            酒店的大樓比較老   N   
95                        可惜美中不足的是浴缸裡有頭髮，可能需要再加強一下房間清潔   N   
96                            房間內無線網路不穩定，常常有2格訊號但卻無法上網   N   
97                                           不過二人房間偏小。   N   
98                        惟酒店較舊，沒有企身浴缸，浴缸太高，不適合同行小童及長者   N   
99             房间光线可以再亮一些，另外，淋浴热水需要调节冷热两个旋钮，感觉有点不方便。     N   
100                      只能开窗子，然而窗子又小，風都不怎么能吹进来又有很大的噪音   N   
101                                         缺点是充电插座太少。   N   
102                                         床及枕頭太軟不太習慣   N   
103                  设备老旧，尤其卫生间内没有独立淋浴，需要爬进浴缸，不适合行动不便者   N   
104                                      服务有待提升，无线很不稳定   N   
105                                隔音加強，減少干擾，停車空間，稍嫌不足   N   
106                                       空調差了點 睡一晚都是塵   N   
107                                      酒店設施比較舊，尤其是廁所   N   
108                                        電視機的接收效果也不好   N   
109                      看到我們是背包客,前台服務員甚至懶得接待我們,自顧自的做事   N   
110                        房間並不像之前評論里說的那樣好,也不像酒店網站上的照片   N   
111                                      房間建議是讓電源插座靠近床   N   
112                                          自助早餐不是很壯觀   N   
113                                    酒店本身很舊需要重新裝修一下了   N   
114                                            不要指望游泳池   N   
115                                             裝飾有點過時   N   
116                                     漏水和一個吹風機還是在地面上   N   
117                                     這絕對根本不應該發生這樣的事   N   
118                                            房間很小很擁擠   N   
119                                      有點怪味兒的健身房設施有限   N   

                                                 token  
0                地點 離 捷運 站 近 ， 要 去 台北 景點 都 方便 服務 人員 熱心  
1          有 問題 也 都 能 即時 解答 房間 乾淨 明亮 ， 設備 俱全 ， 很 安靜 舒適  
2                          這次 已經 是

In [7]:
df_pos=df[df['tag']=="P"]
df_neg=df[df["tag"]=="N"]
posWords = list(df_pos.token)
negWords = list(df_pos.token)

In [9]:
# 加註情感標記
pos_set=[]
neg_set=[]
for d in range(len(df.token)):
    if df.iloc[d]['tag']=="P":
        pos_set.append((df.iloc[d]['token'],'positive'))
    else:
        neg_set.append((df.iloc[d]['token'],'negative'))

In [11]:
print(pos_set[:6])
print(neg_set[:6])

[('地點 離 捷運 站 近 ， 要 去 台北 景點 都 方便 服務 人員 熱心', 'positive'), ('有 問題 也 都 能 即時 解答 房間 乾淨 明亮 ， 設備 俱全 ， 很 安靜 舒適', 'positive'), ('這次 已經 是 第二次 入住 ， 服務 比 上次 更佳', 'positive'), ('職員 對 旅客 有 賓至如歸 的 感覺', 'positive'), ('很 寬敞 ， 很 乾 凈 ， 也 很 安靜 。 我們 住 的 很 舒服 ， 也 很 放鬆 。 酒店 的 早餐 也 很 不錯 。', 'positive'), ('離 火車站 非常 近 ， 交通 又 便利', 'positive')]
[('房間 有 一些 霉味 ！   希望 不會 再 遇上 ！', 'negative'), ('房間 殘舊 ， 清潔 很差 ， 床有 頭髮', 'negative'), ('成間 房 很 臭 ， 好似 舊 酒樓 的 味道', 'negative'), ('以後 都 不會 入住 此 酒店 ！', 'negative'), ('我 不滿 的 除了 他們 待客 態度 外 ， 還有 他們 對 工作 不 認真 ， 沒有 檢查 文件 就 發給 客人 ， 還 反問 那 𥚃 出錯', 'negative'), ('酒店設施 太老舊 ， 房間 太小', 'negative')]


In [12]:
# 開始訓練
# ngrams 函數，以多個字元為分析單位
def ngrams(input, min, max):
    input = input.split(' ')
    output = []
    texts=''  
    for c in range(min,max+1):
        n=c
        for i in range(len(input)-n+1):
            texts=' '.join(input[i:i+n])
            output.append(texts)
    return output

In [13]:
train_set = []
# 把測試組的所有資料叫進來，含字詞及情緒標記 #######
for (words, sentiment) in pos_set + neg_set:
    # 所有的字詞 與 情緒標記
    word_list=ngrams(words, 1,1) 
    
    # 去除停用字 stopwords #####################如果文本量夠多，這些不重要的字就不會出現在規則中
    stopwords=['我', '很', '他', '的', '也' ,"內容"]   
    word_list=[e for e  in word_list if e not in stopwords] 
   
    # 字長控制
    words_filtered = [e for e in word_list if len(e) >= 1]        
    train_set.append((words_filtered, sentiment))

In [14]:
import nltk
# 取得文本中所有的字詞 list，有重覆
def get_words_in_train(train_set):
    all_words = []
    for (words, sentiment) in train_set:
        all_words.extend(words)
    return all_words
# 取得字詞列表，不重復 dictionary_key
def get_word_features(wordlist):
    wordlist = nltk.FreqDist(wordlist)
    word_features = wordlist.keys()
    return word_features
# 字詞是否(True or False)在文件有被引用，dictionary
def extract_features(train_set):
    document_words = set(train_set)
    features = {}
    for word in word_features:
        features['contains(%s)' % word] = (word in document_words)
    return features

In [15]:
# 分類器
# 字詞列表， dictionary_key 不重復
word_features = get_word_features(get_words_in_train(train_set))

# 文字袋比對，每一筆資料，字詞的有無，注意引數 extract_features, tweets　
training = nltk.classify.apply_features(extract_features, train_set)

# 這要花時間 ############
# 找出潛在的規則
classifier = nltk.NaiveBayesClassifier.train(training)

In [16]:
# 列出最重要的 50 條規則
print (classifier.show_most_informative_features(50))

Most Informative Features
            contains(服務) = True           positi : negati =      9.7 : 1.0
             contains(好) = True           positi : negati =      8.3 : 1.0
            contains(沒有) = True           negati : positi =      5.0 : 1.0
            contains(人員) = True           positi : negati =      3.7 : 1.0
             contains(有) = True           positi : negati =      3.4 : 1.0
            contains(乾淨) = True           positi : negati =      3.0 : 1.0
             contains(再) = True           negati : positi =      2.3 : 1.0
            contains(自助) = True           negati : positi =      2.3 : 1.0
             contains(！) = True           negati : positi =      2.3 : 1.0
            contains(覺得) = True           negati : positi =      2.3 : 1.0
             contains(、) = True           positi : negati =      2.3 : 1.0
            contains(設施) = True           positi : negati =      2.3 : 1.0
            contains(還有) = True           negati : positi =      2.3 : 1.0

In [18]:
# save classifier
import pickle
save_classifier = open(path+"naivebayes_test.pickle","wb")
pickle.dump(classifier, save_classifier)
save_classifier.close()

In [19]:
# load classifier
####################################################################################################
import pickle
classifier_f = open(path+"naivebayes_test.pickle", "rb")
classifier = pickle.load(classifier_f)
classifier_f.close()

In [20]:
# 讀取文本資料
#####################################################################
import pandas as pd
df1=pd.read_excel(path+"天成飯店評論斷詞.xlsx", index_col=None)
del df1['Unnamed: 0']
df1.head()

Title                                            Reviews  \
0          很讚  謝謝天成大飯店的全體同仁，因為你們的努力，才有那麼好的住宿體驗！地點方便、服務人員又專業且體...   
1        地點方便  地點離捷運站近，要去台北景點都方便服務人員熱心，有問題也都能即時解答房間乾淨明亮，設備俱全，...   
2    服務人員態度超好  離火車站、捷運很近櫃檯人員都很親切的幫忙解決問題一個人出來住到這裡真是太好了有機會一定會再來...   
3  推薦大家一家好飯店！  飯店到台北車站（火車、捷運、高鐵）非常便利，只要到台北出差都會選擇天成大飯店。  這是為什麼...   
4        休息一下  在臺北過夜一宿，環境舒適，服務人員相當棒，可以為您服務，住在這裏很方便，搭配捷運線想去哪都很...   

                                               token  Rating     Place  \
0  謝謝 天成 大飯店 的 全體同仁 因為 你們 的 努力 才 有 那麼 好 的 住宿 體驗 地...      50  no_place   
1  地點 離 捷運 站 近 要 去 台北 景點 都 方便 服務 人員 熱心 有 問題 也 都 能...      50  no_place   
2  離 火車站 捷運 很近 櫃檯 人員 都 很 親切 的 幫忙 解決問題 一個 人 出來 住 到...      50  no_place   
3  飯店 到 台北 車站 火車 捷運 高鐵 非常 便利 只要 到 台北 出差 都 會 選擇 天成...      50      台灣台北   
4  在 臺北 過夜 一宿 環境 舒適 服務 人員 相當 棒 可以 為 您 服務 住 在 這裏 很...      50     台灣中正區   

  Total_Post        Vote Date_of_Stay  
0      1 篇投稿  no_helpful     2019年10月  
1      1 篇投稿  no_helpful     2019年10月  
2      3 篇投稿  no_helpful     2019年10月  
3     10 篇投稿  no_helpful     2019年10月  
4      1 篇投稿  no_helpful     2019年10月

In [21]:
#刪除空白欄位
delete=[]
for n in range(len(df1)):
    try:
        if (type(df1.token[n])!= str) or (len(df1.token)<1):
            delete.append(n)
    except:
        delete.append(n)
print (len(delete))     
df1=df1.drop(delete)
df1.index=range(len(df1))

0


#  標的檔

In [22]:
# 標的檔
token=df1.token

In [23]:
# 預測
# 正確率計算
yes=0
pred=[]
for t in range(len(token)):
    prediction= classifier.classify(extract_features(token[t].split()))
    pred.append(prediction)

In [24]:
df1['pred']=pred
print ("positive ",len(df1[df1.pred=="positive"]))
print ("negative ",len(df1[df1.pred=="negative"]))

positive  1403
negative  82


# 隨機選取

In [25]:
# 隨機選取 25
import random
num=range(len(df1))
ran=random.sample(num, 25)
df1.iloc[ran][['token','pred']]

token      pred
854   酒店 附近 就 有 機場 大巴 的 車站 , 非常 方便 價錢 也 實惠 可以 從 酒店 出...  positive
1154  方便 的 存取 和 安靜 和 乾淨 的 房間   這是 我 第二次 住 這家 酒店   工作...  positive
31    天成 大飯店 於 上 世紀 七十年代 開業 是 一間 具有 四十多年 歷史 的 四星級 飯店...  positive
110   房間 乾淨 整潔 可以 眺望 繁忙 的 台北 車站 及 車流 搭 捷運 非常 的 方便 飯店...  positive
919   酒店 的 位置 是 無可挑剔 的   他們 甚至 還有 自己 的 地鐵站 出口 3 米   ...  positive
872   從 桃園 機場 坐 巴士 下車 , 走 一會兒 就 到 了 酒店 就 在 台北 站 附近 ,...  positive
759   地理位置 超好 去 各种 地方 都 很 便利 虽然 在 车站 周围 但 一点 也 不会 觉得...  negative
1361  10.1   -   10.6 住 5 晚 在 這家 酒店 感覺 就 像 呆 在 家裏 , ...  positive
239   不錯 設施 齊全 乾淨 服務員 態度 好 地點 交通 太 方便 去 那 都 有 各種 交通工...  positive
571   房間 整潔 乾淨 服務員 態度 和善 位於 台北 車站 旁 交通 便利 周圍 機 生活 能 ...  positive
436   登記 前臺 飯店 服務員 服務中心 和 早 餐廳 前線 員 全部 也 是 服務 慇勤 笑容 ...  positive
855   酒店 的 位置 很 好 但是 我 這次 住 的 是 15 樓 台北 站 那 一側 的 房間 ...  positive
767   服务 有待 提升 无线 很 不 稳定 整体 还 算 可以 早餐 不错 地理位置 很 是 ok...  negative
272   房間 舒適 整潔 3 人 房間 尚算 寬敞 位置 就 在 台北 車站 側 機場 巴士站 亦 ...  positive
1367  這是 我 第二次 住 在 這家 jotel   一個 很棒 的 位置 但是 價格 在 臺北 ...  positive
870   我們 是 看 了 論壇 上 的 評論 之後 才 在 預訂 的 台北 天成 飯店 飯店 所在 ...  positive
629   我们 订 的 是 家庭 房 两个 房间 虽然 不大 但是 非常 干净 整洁 按摩 浴缸 超级...  negative
318   近 火車站 周圍 也 有 很多 逛點 尤其 是 我 父親 不好 入眠 也 說 睡 的 很 好...  positive
1061  我們 第一次 在 臺北 住 了 幾個 晚上 我們 去 了 今年 在 臺北 天成 大酒店 因爲...  positive
903   前臺 的 員工 都 非常 善良     他們 很 友好 的 預訂 保留 他們 在 一個 我們...  positive
446   前台 禮賓 部及 管家 部 員工 笑容 親切 親切 窩心 無微不至 值得 入住 酒店 惟 酒...  negative
840   從 地鐵 一出 站 , 走 1 分鐘 就 到 飯店 了   而且 離得 最近 的 那個 出口...  positive
365   交通 方便 由 機場 到 酒店 只 需 45 分鐘 員工 態度 友善 房間 整齊清潔 設備齊...  positive
736   位處 台北 捷運 Ｍ ３ 出口 地理位置 便利 服務 人員 親切 房間 狀態 良好 出遊 經...  positive
505   綜合 評比 了 多家 飯店 后 最終 選擇 的 天成 入住 主要 是 看 它 的 位置 市中...  positive

In [26]:
#人工判斷
pred=['negative','positive','negative','positive','positive','positive','negative','negative','positive','positive','positive','negative','negative','positive','positive','positive','positive','positive','positive','negative','negative','negative','positive','positive','positive']
judge=['negative','positive','positive','positive','positive','positive','negative','negative','positive','positive','negative','positive','negative','positive','positive','negative','positive','positive','positive','positive','positive','positive','positive','positive','positive']
#正確率計算
yes=0
for n in range(len(judge)):
    if pred[n] == judge[n]:
        yes+=1
acc=yes/len(judge)
print (acc)

0.72


In [28]:
# 檢查某一筆資料是否正確
print (df1.token[99])
print (df1.pred[99])

來 台灣 旅遊 幾個 月 前 就 定 了 天成 飯店 交通 非常 便利 台北 車站 旁邊 去 各個 地方 都 很 方便 酒店 服務 很 好 哦 非常 熱情 …   房間 很 乾 凈 整潔 … 很 是 滿意 早餐 也 都 非常 好以 後來 台北 還住 這裏 推薦 哦 ⭐ ️ ⭐ ️ ⭐ ️ ⭐ ️ ⭐ ️
positive


In [30]:
df讚美=pd.read_excel(path+"讚美詞.xlsx",index_col=None)
pos=set(list(df讚美.positive))
df譴責=pd.read_excel(path+"譴責詞.xlsx",index_col=None)
neg=set(list(df譴責.negative))

In [31]:
df1

Title                                            Reviews  \
0                       很讚  謝謝天成大飯店的全體同仁，因為你們的努力，才有那麼好的住宿體驗！地點方便、服務人員又專業且體...   
1                     地點方便  地點離捷運站近，要去台北景點都方便服務人員熱心，有問題也都能即時解答房間乾淨明亮，設備俱全，...   
2                 服務人員態度超好  離火車站、捷運很近櫃檯人員都很親切的幫忙解決問題一個人出來住到這裡真是太好了有機會一定會再來...   
3               推薦大家一家好飯店！  飯店到台北車站（火車、捷運、高鐵）非常便利，只要到台北出差都會選擇天成大飯店。  這是為什麼...   
4                     休息一下  在臺北過夜一宿，環境舒適，服務人員相當棒，可以為您服務，住在這裏很方便，搭配捷運線想去哪都很...   
5                 颱風來襲也不用怕  天成是台北車站附近的最佳飯店, 不用帶雨傘也可到達車站, 還有便利店在樓下, 完全不用擔心不...   
6                  一流的台北洒店  這次已經是第二次入住，服務比上次更佳。職員對旅客有賓至如歸的感覺。在填寫房間意見調查後，職員...   
7                     飯店很美  整體飯店氣氛很好，對於我的請求都細心的為我處理好了，雖然飯店的歷史長遠，但我想這大概是為什麼...   
8                   很不錯的酒店  我們在這家酒店住了4天3晚。我們的房間是初級套房。很寬敞，很乾凈，也很安靜。我們住的很舒服，...   
9                   高於我的預期  這家酒店靠近各大交通站台：列車站，巴士站、火車站。。。等等。酒店住宿還包早餐，早餐高於我的預...   
10                 地點位置非常好  今年出差還是住在這裡，交通地點位置真的是無話可說的棒，服務人員服務品質還是維持著高水準，真是...   
11                  交通超方面!  飯店開了已有點日子了，裝修不屬型格潮款屬别，飯店位於台北捷運M3出入口不遠處，前往台北火車站...   
12                  房間有霉味！  今次入住了３晚909號雙床有景觀房，房間有一些霉味！把空調連續24小時開放…直到第三晚回到酒...   
13                  台北車站很近  雖然設施有點舊，但是還算乾淨。最好的一點是距離台北車站很近。如果要在台北車站趕早班車或者晚班...   
14                 台北最好的飯店  交通便利、服務人員親切、房間乾淨、床很舒服、早餐好吃、價錢合理、CP值高,下次還會想來住，無...   
15           2019年在台灣度過的假期  台北天成大飯店位置方便，毗鄰台北車站。酒店的服務團隊非常棒，非常熱情好客。自助早餐很好吃，有...   
16                    不錯選擇  很久車站，及地下街，遊台北很方便，洒店較舊，但房間尚可，早餐合格，但整體整潔有代改善，員工服...   
17                     地點好  老字號飯店，有一種陳舊感，但地點非常好，四通八達，房間也不會太小，應有盡有，櫃檯人員親切，雖...   
18                 非常方便的酒店  位於台北火車站旁的洒店，交通非常方便，價格並不貴，服務人員主動有禮，房間不太大，而且比較舊式...   
19                GOOOOOOD  交通便利，櫃檯人員笑容甜美。 當有任何問題詢問飯店人員時都很積極協助幫忙，真的非常感謝入住期...   
20                    貼心服務  其中一位服務員Meg她的服務令我印象深刻，她馬上發現了我的需求立即為我處理，讓我感覺備受尊重...   
21    Cosmos Hotel Taipei   大廳氣氛良好  房間內也很舒服😌 重點是服務人員態度很好 如果下次來台北的話 也要住這邊 交...   
22                  台北最適住宿  台北住宿的第一選擇，交通便利，環境簡潔、優雅，早餐豐盛，服務親切、貼心，不論是旅遊或商務皆是...   
23                   親切的服務  我們這個月有到天成飯店的百合西餐廳用餐過，氣氛很不錯喔，餐飲好吃、親切服務很棒！很推薦大家參...   
24            good service  很滿意 房間，也很讚賞員工的禮貌及服務態度。下次一定會回來住。 每一次回來所有員工包括房務部...   
25                地方大 環境一般  地方大，既能完全打開行李，也有桌子和化妝枱，浴室內的浴缸也不少！不過在走廊、升降機或大堂都隱...   
26                乾淨，地點很方便  入住時適逢是我的生日，我收到他們的蛋糕及職員嘅簽名，感覺很溫馨很感動！  可是早餐欠缺新意，...   
27          房間殘舊，清潔很差，有異味！  房間殘舊，清潔很差，床有頭髮，成間房很臭，好似舊酒樓的味道。浴缸很多污垢。由於有免費送一杯飲...   
28                     位置佳  屬老牌酒店，服務態度不曾。酒店與車站相連，可直達地下街及捷運，附近又有不少食店與商店，生活機...   
29               櫃台同仁都很有耐心  櫃台同仁都非常優秀 辦理手續時是非常開心的，房間隔音差 ，但櫃台非常有耐心的幫我們處理，很棒...   
...                    ...                                                ...   
1455                  客房太冷  是陳舊但不錯的酒店, 但最糟糕的是房間在這樣一個即使在雙城！ 無法想象一下就像將在一月份的人...   
1456                  2晚住宿  良好的位置, abt 臺北車站5分鐘即可到達火車站和1（購物中心）到西門町。   房間很小但...   
1457            非常適閤家庭帶着孩子  我們住在這裏的期間2N 10days 臺灣旅程。 房間很小但是傢俱以適合我們的行李和排列。 ...   
1458                  便捷舒適  我們兩次在宇宙酒店的總數爲7晚和與它很好。 先上我們的入住體驗我們的行政樓層的房間像一個第5...   
1459                  臺北中心  我的第二次來, 服務標準有一位不丟棄。 價格合理, 位置很好, 舒適的客房。 一種真正的很好...   
1460        位置絕佳的員工樂於助人的員工  我剛從1周昨天麗悅酒店住宿。 其主要的賣點是它的整個過程由於位置的便利。 我想要一些鑄造注意...   
1461            很不錯的酒店位置優越  確定, 此酒店也是臺北車站這樣如果您計劃讓都市那麼這是您絕好的位置。 服務非常好, 房間是乾...   
1462       酒店的位置很棒的寬敞豪華四人間  我們（系列2A 十幾歲男孩和2）住在宇宙2014年11月13日 9 - 從。 已選擇直接從酒...   
1463               房間位置超棒！  我預訂的是最棒的四人間。 還有我的嬰兒與我女朋友。 我這的招待我們的男包的一個重頭戲。  我...   
1464                  住的很棒  住的很好的火車站, 地鐵站, 購物中心。 方便到達所有名勝古蹟。 客房中, 這是酒店設計的全...   
1465             很好的酒店位置優越  酒店都不用其中一個從臺北車站可以趕上火車正常, 高速火車或地鐵的郊區。 將總綫從機場可以立馬...   
1466                 小小的抱怨  臺灣我們很喜歡, 我們很喜歡我們在臺北住宿, 也可能因爲服務由宇宙酒店是非常完美！ 您將享受...   
1467        無法 nk 精簡的另一個地方  酒店位於優越的地理位置與較好的團隊  服務 - 酒店擁有適當的團隊成員選擇。 每個人都要有禮...   
1468                    同去  位置是這個件很好的事情。 兩臺北鐵路這是巨大的且易於丟失除非您注意中的退出號碼... 緊鄰許...   
1469              位置優越的好酒店  對於喜歡臺灣果然, 臺北車站要保持的地方。 宇宙酒店位於退出 M3地鐵站的。 地鐵站在此, ...   
1470        不錯的住宿酒店臺北的海外遊客  它是實際&100％從宇宙酒店的承諾服務, &我真的感謝所有的員工, 並希望他們的所有&將明年...   
1471                住宿的好地方  它是一個很好的選擇住的酒店在宇宙, 尤其是地理位置與它。 它接近 HSR 附近的火車站和客運...   
1472                  位置優越  完美在臺北期間我的時間。 酒店的位置是最好的屬性

In [32]:
n_pos=[]
n_neg=[]
for n in range(len(df1)):
    set1=set(df1.iloc[n]['token'].split())
    num=set1.intersection(pos)
    n_pos.append(num)
    num=set1.intersection(neg)
    n_neg.append(num)      

In [33]:
df1['pos']=n_pos
df1["neg"]=n_neg

In [34]:
key=['飯店','人員','房間','位置']
n_key=[]
for n in range(len(df1)):
    set1=set(df1.iloc[n]['token'].split())
    num=set1.intersection(key)
    n_key.append(num)
find=[]
for s in range(len(df1)):     
    amount=0 
    for k in key:         
        count=df1.token[s].count(k,0)
        amount+=count
    find.append(amount) 
df1['key_num']=find
df1["key"]=n_key

In [36]:
# 選取部分欄位，製作新 dataframe
df2=df1[['Title','token','pos', 'neg','pred','key']]
mask=df1["key_num"]>0
df_hotel=df2[mask] 
df_hotel

Title                                              token  \
0                       很讚  謝謝 天成 大飯店 的 全體同仁 因為 你們 的 努力 才 有 那麼 好 的 住宿 體驗 地...   
1                     地點方便  地點 離 捷運 站 近 要 去 台北 景點 都 方便 服務 人員 熱心 有 問題 也 都 能...   
2                 服務人員態度超好  離 火車站 捷運 很近 櫃檯 人員 都 很 親切 的 幫忙 解決問題 一個 人 出來 住 到...   
3               推薦大家一家好飯店！  飯店 到 台北 車站 火車 捷運 高鐵 非常 便利 只要 到 台北 出差 都 會 選擇 天成...   
4                     休息一下  在 臺北 過夜 一宿 環境 舒適 服務 人員 相當 棒 可以 為 您 服務 住 在 這裏 很...   
5                 颱風來襲也不用怕  天成 是 台北 車站 附近 的 最佳 飯店 ,   不用 帶 雨傘 也 可 到達 車站 , ...   
6                  一流的台北洒店  這次 已經 是 第二次 入住 服務 比 上次 更佳 職員 對 旅客 有 賓至如歸 的 感覺 ...   
7                     飯店很美  整體 飯店 氣氛 很 好 對於 我 的 請求 都 細心 的 為 我 處理 好 了 雖然 飯店...   
8                   很不錯的酒店  我們 在 這家 酒店 住 了 4 天 3 晚 我們 的 房間 是 初級 套房 很 寬敞 很 ...   
10                 地點位置非常好  今年 出差 還是 住 在 這裡 交通 地點 位置 真的 是 無話可說 的 棒 服務 人員 服...   
11                  交通超方面!  飯店 開 了 已 有點 日子 了 裝修 不屬型 格潮款 屬别 飯店 位於 台北 捷運 M3 ...   
12                  房間有霉味！  今 次 入住 了 ３ 晚 909 號雙床 有 景觀 房 房間 有 一些 霉味 把 空調 連續...   
14                 台北最好的飯店  交通 便利 服務 人員 親切 房間 乾淨 床 很 舒服 早餐 好吃 價錢 合理 CP 值高 ...   
15           2019年在台灣度過的假期  台北 天成 大飯店 位置 方便 毗鄰 台北 車站 酒店 的 服務 團隊 非常 棒 非常 熱情...   
16                    不錯選擇  很 久 車站 及 地下街 遊 台北 很 方便 洒店 較 舊 但 房間 尚可 早餐 合格 但 ...   
17                     地點好  老字號 飯店 有 一種 陳舊 感 但 地點 非常 好 四通八達 房間 也 不會 太小 應有盡...   
18                 非常方便的酒店  位於 台北 火車站 旁 的 洒 店 交通 非常 方便 價格 並不貴 服務 人員 主動 有禮 ...   
19                GOOOOOOD  交通 便利 櫃檯 人員 笑容 甜美   當有 任何 問題 詢問 飯店 人員 時 都 很 積極...   
21    Cosmos Hotel Taipei   大廳 氣氛 良好     房間內 也 很 舒服   重點 是 服務 人員 態度 很 好   ...   
22                  台北最適住宿  台北 住宿 的 第一 選擇 交通 便利 環境 簡潔 優雅 早餐 豐盛 服務 親切 貼心 不論...   
23                   親切的服務  我們 這個 月 有 到 天成 飯店 的 百合 西餐廳 用餐 過 氣氛 很 不錯 喔 餐飲 好...   
24            good service  很 滿意   房間 也 很 讚賞 員工 的 禮貌 及 服務態度 下次 一定 會 回來 住  ...   
27          房間殘舊，清潔很差，有異味！  房間 殘舊 清潔 很 差床 有 頭髮 成間 房 很 臭 好似 舊 酒樓 的 味道 浴缸 很多...   
29               櫃台同仁都很有耐心  櫃台 同仁 都 非常 優秀   辦理手續 時 是 非常 開心 的 房間 隔音 差   但櫃台...   
30               位置方便的住宿地點  這件 飯店 已經 很多年 了 小時候 曾經 住過 一次 這次 帶著 爸媽 入住 三人房 因為...   
31            一間歷史悠久的四星級酒店  天成 大飯店 於 上 世紀 七十年代 開業 是 一間 具有 四十多年 歷史 的 四星級 飯店...   
32              地理位置和服務都很棒  這個 位置 無與倫比 離 台北 站 非常 近 附近 有 許多 商店 餐廳 麵 包店 和 咖啡...   
34               位置優越的天成飯店  很久没 來 天成 飯店 了 今天 再次 拜訪 很 喜歡 大廳 乾淨 高級 的 歐洲 風味 因...   
35               舒適溫馨的天成飯店  今天 和 媽媽 來 天成 飯店 很 喜歡 這家 飯店 溫馨 又 典雅 的 妝 潢 飯店 服務...   
36                  整體品質絕佳  位 在 台北 車站 旁 交通 非常 方便 甚至 不用 一分鐘 就 可 走 到 捷運 3 號 ...   
...                    ...                                                ...   
1452              容易被遺忘的酒店  此 酒店 非常 方便 許多 旅客 對 它 位於 臺北 車站   的 以外 ,   請 爲此 ...   
1453          亞裔員工小大小相稱的客房  抵達 上 ,   我們 的 出租車 過濾 到 空間 中 的 酒店 大堂 附近 的 酒店 和 ...   
1454          鞋盒大男孩位置很棒的服務  在 2014 年 12 月 住 2 晚 的   位置 很 好 旁邊 臺北 火車站 M3 退出...   
1455                  客房太冷  是 陳舊 但 不錯 的 酒店 ,   但 最 糟糕 的 是 房間 在 這樣 一個 即使 在 ...   
1456                  2晚住宿  良好 的 位置 ,   abt   臺北 車站 5 分鐘 即可 到達 火車站 和 1 購物中...   
1457            非常適閤家庭帶着孩子  我們 住 在 這裏 的 期間 2N   10days   臺灣 旅程   房間 很小 但是 ...   
1458                  便捷舒適  我們 兩次 在 宇宙 酒店 的 總數 爲 7 晚 和 與 它 很 好   先 上 我們 的 ...   
1459                  臺北中心  我 的 第二次 來 ,   服務 標準 有 一位 不 丟棄   價格合理 ,   位置 很 ...   
1460        位置絕佳的員工樂於助人的員工  我剛 從 1 周 昨天 麗悅 酒店 住宿   其 主要 的 賣點 是 它 的 整個 過程 由...   
1461            很不錯的酒店位置優越  確定 ,   此 酒店 也 是 臺北 車站 這樣 如果 您 計劃 讓 都市 那麼 這是 您 ...   
1462       酒店的位置很棒的寬敞豪華四人間  我們 系列 2A   十幾歲 男孩 和 2 住 在 宇宙 2014 年 11 月 13 日 ...   
1463               房間位置超棒！  我 預訂 的 是 最 棒 的 四 人間   還有 我 的 嬰兒 與 我 女朋友   我 這的...   
1464                  住的很棒  住 的 很 好 的 火車站 ,   地鐵站 ,   購物中心   方便 到達 所有 名勝古蹟...   
1465             很好的酒店位置優越  酒店 都 不用 其中 一個 從 臺北 車站 可以 趕上 火車 正常 ,   高速 火車 或 ...   
1466                 小小的抱怨  臺灣 我們 很 喜歡 ,   我們 很 喜歡 我們 在 臺北 住宿 ,   也 可能 因爲 ...   
1467        無法 nk 精簡的另一個地方  酒店 位於 優越 的 地理位置 與 較 好 的 團隊     服務   -   酒店 擁有 ...   
1468                    同去  位置 是 這個 件 很 好 的 事情   兩 臺北 鐵路 這是 巨大 的 且 易於 丟失 除...   
1469              位置優越的好酒店  對於 喜歡 臺灣 果然 ,   臺北 車站 要 

In [37]:
print ("positive ",len(df_hotel[df_hotel.pred=="positive"]))
print ("negative ",len(df_hotel[df_hotel.pred=="negative"]))

positive  1193
negative  66


## 結論
透過此次作業可以看出飯店評論正負面情緒的文本量相差極大，
positive  1403
negative  82
若以消費者給出的評分(rating)來看，5分=864、4分=528、3分=79、2分=10、1分=4
給出4~5分的量也達1392個，跟從評論分析出正面的量1403其實是相當接近的。當然每則評論其實都包含了正負內容，並非絕對正面或負面，因此用評論的正負情緒結果跟rating結果比較是有些突兀的，但是個人認為這樣的結果還是相當有趣。